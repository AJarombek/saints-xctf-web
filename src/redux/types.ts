/**
 * Types for the Redux store.
 * @author Andrew Jarombek
 * @since 7/23/2020
 */

export type RootState = {
    auth: AuthState,
    logs: LogsState
}

export type LogsState = {
    isFetching: boolean;
    didInvalidate: boolean;
    lastUpdated: number;
    items: Logs;
    feeds: LogFeeds;
    newComments: NewComments;
}

export type AuthState = {
    auth: Auth;
    user: Users;
}

export interface Meta {
    isFetching?: boolean;
    lastUpdated?: number;
    didInvalidate?: boolean;
    serverError?: string;
}

export type Auth = {
    isFetching?: boolean;
    lastUpdated?: number;
    signedInUser?: string;
    status?: string;
}

export type Users = {
    [key: string]: UserMeta;
}

export interface UserMeta extends User, Meta {}

export interface User {
    activation_code?: string;
    class_year?: number;
    deleted?: string;
    description?: string;
    email?: string;
    favorite_event?: string;
    first?: string;
    last?: string;
    last_signin?: string;
    location?: string;
    member_since?: string;
    password?: string;
    salt?: string;
    subscribed?: string;
    username?: string;
    week_start?: string;
}

enum LogType { RUN = "run", BIKE = "bike", SWIM = "swim", OTHER = "other" }
enum Metric { MILES = "miles", KILOMETERS = "kilometers", METERS = "meters" }

export type Log = {
    log_id: number;
    username: string;
    first: string;
    last: string;
    name: string;
    date: string;
    type: LogType;
    distance?: number;
    metric?: Metric;
    miles?: number;
    time?: string;
    pace?: string;
    feel: number;
    location?: string;
    description?: string;
    timeCreated?: string;
    comments?: Comment[];
}

export type Logs = {
    [key: string]: Record<string, Log>;
}

export type LogFeeds = {
    [key: string]: {
        filterBy: string;
        bucket: string;
        pages: LogFeedPages
    };
}

export type LogFeedPages = {
    [key: string]: LogFeedPage
}

export type LogFeedPage = {
    isFetching: boolean;
    lastUpdated: number;
    items: Log[];
    serverError: string;
}

export type Comment = {
    comment_id: number;
    username: string;
    first: string;
    last: string;
    log_id: number;
    time: string;
    content: string;
}

export type NewComments = {
    [key: string]: {
        isFetching: boolean;
        lastUpdated: number;
        created?: boolean;
        serverError?: string;
    }
}
