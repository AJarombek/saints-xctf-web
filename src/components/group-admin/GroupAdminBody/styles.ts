/**
 * JSS styles for the GroupAdminBody component.
 * @author Andrew Jarombek
 * @since 1/16/2021
 */

import Mixins, { AJComponentMixins, FontMixins } from '../../../styles/mixins';

export default {
  groupAdminBody: {
    ...Mixins.profileAndGroupBody(),

    '& > aside > .tabs': {
      width: 200
    }
  },
  mobileTabs: {
    display: 'none'
  },
  select: {
    ...AJComponentMixins.ajSelect()
  },
  title: {
    ...FontMixins.robotoSlabBold(),
    fontSize: 20,
    marginTop: 30
  },
  subTitle: {
    ...FontMixins.robotoBold(),
    marginTop: 15,
    color: '#777'
  },
  text: {
    marginLeft: 15
  },
  '@media screen and (max-width: 1200px)': {
    text: {
      marginLeft: 10
    }
  },
  '@media screen and (max-width: 900px)': {
    groupAdminBody: {
      '& > section': {
        margin: 0
      }
    }
  },
  '@media screen and (max-width: 750px)': {
    groupAdminBody: {
      flexDirection: 'column',

      '& > aside': {
        display: 'flex'
      },

      '& > div:nth-child(1)': {
        flexBasis: '30%'
      },

      '& .tabs': {
        display: 'none'
      }
    },
    mobileTabs: {
      ...Mixins.profileAndGroupMobileTabs(),
      margin: '0 0 0 30px'
    },
    select: {
      width: '100%'
    }
  }
};
