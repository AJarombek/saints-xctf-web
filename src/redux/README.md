### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
