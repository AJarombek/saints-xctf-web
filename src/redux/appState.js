/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
