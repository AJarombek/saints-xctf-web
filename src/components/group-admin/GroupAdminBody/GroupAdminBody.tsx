/**
 * GroupAdminBody component which allows admins to manage users, notifications, and group information.
 * @author Andrew Jarombek
 * @since 1/16/2021
 */

import React, { useEffect, useMemo, useState } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import { GroupMeta, RootState, TeamMeta } from '../../../redux/types';
import PageTabs from '../../shared/PageTabs';
import { useDispatch, useSelector } from 'react-redux';
import { getGroup, getGroupTeam } from '../../../redux/modules/groups';
import classNames from 'classnames';
import ManageUsers from '../ManageUsers/ManageUsers';
import SendActivationCode from '../SendActivationCode';
import EditGroup from '../EditGroup';
import { AJSelect } from 'jarombek-react-components';
import { Tab } from '../../shared/PageTabs/PageTabs';
import { GroupTab } from '../../group/GroupBody/GroupBody';

interface Props {
  group: GroupMeta;
  groupId: number;
}

export enum GroupAdminTab {
  MANAGE_USERS = 'manage-users',
  SEND_ACTIVATION_CODE = 'send-activation-code',
  EDIT_GROUP = 'edit-group'
}

const useStyles = createUseStyles(styles);

const GroupAdminBody: React.FunctionComponent<Props> = ({ group, groupId }) => {
  const classes = useStyles();

  const dispatch = useDispatch();
  const teamInfo: Record<string, TeamMeta> = useSelector((state: RootState) => state.groups.team ?? {});

  const [tab, setTab] = useState<GroupAdminTab>(GroupAdminTab.MANAGE_USERS);

  useEffect(() => {
    if (!group) {
      dispatch(getGroup(+groupId));
    }
  }, [groupId, group, dispatch]);

  useEffect(() => {
    if (!teamInfo[groupId]) {
      dispatch(getGroupTeam(groupId));
    }
  }, [dispatch, groupId, teamInfo]);

  const teamTitle: string = useMemo(() => {
    return teamInfo[groupId]?.title;
  }, [groupId, teamInfo]);

  const tabs = useMemo(
    () => [
      {
        tab: GroupAdminTab.MANAGE_USERS,
        onView: (): void => setTab(GroupAdminTab.MANAGE_USERS),
        content: 'Manage Users'
      },
      {
        tab: GroupAdminTab.SEND_ACTIVATION_CODE,
        onView: (): void => setTab(GroupAdminTab.SEND_ACTIVATION_CODE),
        content: 'Send Activation Code'
      },
      { tab: GroupAdminTab.EDIT_GROUP, onView: (): void => setTab(GroupAdminTab.EDIT_GROUP), content: 'Edit Group' }
    ],
    []
  );

  return (
    <div className={classes.groupAdminBody}>
      <aside>
        <div>
          <h5 className={classNames(classes.title, classes.text)}>{group?.group_title}</h5>
          <div className={classNames(classes.subTitle, classes.text)}>{teamTitle}</div>
        </div>
        <PageTabs currentTab={tab} tabs={tabs} />
        <div className={classes.mobileTabs}>
          <AJSelect
            options={...tabs.map((tab: Tab) => ({ content: tab.content, value: tab.tab }))}
            defaultOption={1}
            onClickListOption={(item: { content: string; value: string }): void => setTab(item.value as GroupAdminTab)}
            className={classNames(classes.select, 'mobileTabsSelect')}
          />
        </div>
      </aside>
      <section>
        {tab === GroupAdminTab.MANAGE_USERS && <ManageUsers groupId={groupId} />}
        {tab === GroupAdminTab.SEND_ACTIVATION_CODE && <SendActivationCode groupId={groupId} />}
        {tab === GroupAdminTab.EDIT_GROUP && <EditGroup group={group} />}
      </section>
    </div>
  );
};

export default GroupAdminBody;
