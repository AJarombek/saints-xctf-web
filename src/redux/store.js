/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
