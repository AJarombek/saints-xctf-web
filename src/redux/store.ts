/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
