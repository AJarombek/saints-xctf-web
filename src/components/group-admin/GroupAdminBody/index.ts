import GroupAdminBody from './GroupAdminBody';

export default GroupAdminBody;
