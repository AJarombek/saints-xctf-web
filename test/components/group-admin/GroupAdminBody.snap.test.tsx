/**
 * Snapshot test for the GroupAdminBody component.
 * @author Andrew Jarombek
 * @since 3/5/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import { emptyStore } from '../../test-utils/storeMocks';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { alumni } from '../../test-utils/groupMocks';
import GroupAdminBody from '../../../src/components/group-admin/GroupAdminBody';

const mockStore = configureStore([thunk]);

describe('GroupAdminBody Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore(emptyStore);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Provider store={store}>
            <GroupAdminBody group={alumni} groupId={1} />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
