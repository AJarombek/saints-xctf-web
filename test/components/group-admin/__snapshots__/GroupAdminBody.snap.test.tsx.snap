// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupAdminBody Snapshot Tests renders correctly 1`] = `
<div
  className="groupAdminBody-0-1-1"
>
  <aside>
    <h5
      className="title-0-1-2 text-0-1-4"
    >
      Alumni
    </h5>
    <div
      className="subTitle-0-1-3 text-0-1-4"
    />
    <div
      className="tabs-0-1-5 tabs"
    >
      <p
        className="tab-0-1-6 currentTab-0-1-7"
        onClick={[Function]}
      >
        Manage Users
      </p>
      <p
        className="tab-0-1-6"
        onClick={[Function]}
      >
        Send Activation Code
      </p>
      <p
        className="tab-0-1-6"
        onClick={[Function]}
      >
        Edit Group
      </p>
    </div>
  </aside>
  <section>
    <div
      className="manageUsers-0-1-8"
    >
      <h3
        className="title-0-1-9"
      >
        Manage Users
      </h3>
      <div
        className="container-0-1-10"
      />
    </div>
  </section>
</div>
`;
